johnzon-core/src/main/java/org/apache/johnzon/core/Snippet.java

```diff
@@ -0,0 +1,435 @@
+                    generator.flush();
+                }
+            }
+        }
+
+        private void write(final JsonArray array) {
+            if (snippet.terminate()) {
+                return;
+            }
+
+            if (array.isEmpty()) {
+                generator.write(array);
+                generator.flush();
+                return;
+            }
+
+            generator.writeStartArray();
+            generator.flush();
+            for (final JsonValue jsonValue : array) {
+                if (snippet.terminate()) {
+                    break;
+                }
+                write(jsonValue);
+            }
+            generator.writeEnd();
+            generator.flush();
+        }
+
+        private void write(final JsonObject object) {
+            if (snippet.terminate()) {
+                return;
+            }
+
+            if (object.isEmpty()) {
+                generator.write(object);
+                generator.flush();
+                return;
+            }
+
+            generator.writeStartObject();
+            generator.flush();
+            for (final Map.Entry<String, JsonValue> entry : object.entrySet()) {
+                if (snippet.terminate()) {
+                    break;
+                }
+                write(entry.getKey(), entry.getValue());
+            }
+            generator.writeEnd();
+            generator.flush();
+        }
+
+        private void write(final String name, final JsonValue value) {
+            if (snippet.terminate()) {
+                return;
+            }
+
+            switch (value.getValueType()) {
+                case ARRAY:
+                    generator.writeStartArray(name);
+                    generator.flush();
+                    final JsonArray array = value.asJsonArray();
+                    for (final JsonValue jsonValue : array) {
+                        if (snippet.terminate()) {
+                            break;
+                        }
+                        write(jsonValue);
+                    }
+                    generator.writeEnd();
+                    generator.flush();
+
+                    break;
+                case OBJECT:
+                    generator.writeStartObject(name);
+                    generator.flush();
+                    final JsonObject object = value.asJsonObject();
+                    for (final Map.Entry<String, JsonValue> keyval : object.entrySet()) {
+                        if (snippet.terminate()) {
+                            break;
+                        }
+                        write(keyval.getKey(), keyval.getValue());
+                    }
+                    generator.writeEnd();
+                    generator.flush();
+
+                    break;
+                default: {
+                    generator.write(name, value);
+                    generator.flush();
+                }
+            }
+        }
+
+        private String get() {
```

Comment:
your flush issue comes from there, let's drop the flush and call get after the auto-close in https://github.com/apache/johnzon/pull/84/files#diff-f8878dc73ad8ecae625a2bf13a77b5718505c77f547c31cefc2454e42454138aR113 which should be written as:

```
    public static String of(final JsonValue value, final int max) {
        final Snippet snippet = new Snippet(max);
        try (final Snippet s  = snippet) { // use the style you want but the important part is to close before using the in mem stream
             s.of(value)
        }
        return s.get();
    }
```

side note: without that, the generator.close is not called and we leak as mentionned in early comments

Reply:
Flush content covered on the lists.